### PR TITLE
Enhance/META-3919 : Create refresh classification task

### PR DESCRIPTION
**Pull request by sumandas0**
## To reduce the no of task created we have introduced new type of task which will ensure classification is only propagated to reachable assets and deleted from unreachable assets

> We have created a new type of task `CLASSIFICATION_REFRESH_PROPAGATION` which will be scheduled in case of any propagation removal is needed. And if this task is already scheduled for one vertex id and it is in the pending state then we will not schedule any more refresh propagation tasks.

## Type of change
- [x] New feature (adds functionality)

### Development

- [x] Application changes have been tested thoroughly

### CODE REVIEW

File: repository/src/main/java/org/apache/atlas/repository/graph/GraphBackedSearchIndexer.java [hunk 1]
```diff
@@ -386,6 +386,8 @@ private void initialize(AtlasGraph graph) throws RepositoryException, IndexExcep
 
             createCommonVertexIndex(management, TASK_TYPE, UniqueKind.NONE, String.class, SINGLE, true, false, true);
             createCommonVertexIndex(management, TASK_CREATED_BY, UniqueKind.NONE, String.class, SINGLE, false, false, true);
+            createCommonVertexIndex(management, TASK_UNIQUE_PARAMETER, UniqueKind.NONE, String.class, SINGLE, true, false, false);
```
Comment (mehtaanshul) [inline repository/src/main/java/org/apache/atlas/repository/graph/GraphBackedSearchIndexer.java:389]: make isStringField: true for both

**Comment by mehtaanshul** [comment]
What will happen to the existing DELETE_ONLY_PROPAGATION tasks, will those run successfully?

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v1/DeleteHandlerV1.java [hunk 2]
```diff
@@ -1281,61 +1286,50 @@ private void setBlockedClassificationIds(AtlasEdge edge, List<String> classifica
...
-            return;
-        }
-
-        Map<String, Object> taskParams  = ClassificationTask.toParameters(deletedEdgeId, classificationVertexId);
-        AtlasTask           task        = taskManagement.createTask(taskType, currentUser, taskParams);
-
-        RequestContext.get().queueTask(task);
-    }
-
     public void createAndQueueTask(String taskType, AtlasEdge relationshipEdge, AtlasRelationship relationship) {
```
Comment (mehtaanshul) [inline repository/src/main/java/org/apache/atlas/repository/store/graph/v1/DeleteHandlerV1.java:1321]: CLASSIFICATION_PROPAGATION_RELATIONSHIP_UPDATE - Want to discuss on unique and reference id here

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v1/DeleteHandlerV1.java [hunk 3]
```diff
@@ -1351,23 +1345,51 @@ public void createAndQueueTask(String taskType, AtlasEdge edge) {
...
+                LOG.info("Task is already scheduled for classification id {}, no need to schedule task for edge {}", currentClassificationId, edge.getIdForDisplay());
+                continue;
             }
 
-            Map<String, Object> taskParams = ClassificationTask.toParameters(currentClassificationVertex.getIdForDisplay(),
-                    referenceVertex.getIdForDisplay(), isTermEntityEdge);
-            AtlasTask task  =  taskManagement.createTask(taskType,currentUser,taskParams);
+            boolean removePropagationOnEntityDelete = GraphHelper.getRemovePropagations(currentClassificationVertex);
+
+            if (!(isTermEntityEdge || removePropagationOnEntityDelete)) {
```
Comment (mehtaanshul) [inline repository/src/main/java/org/apache/atlas/repository/store/graph/v1/DeleteHandlerV1.java:1367]: Discuss this

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v2/AtlasEntityStoreV2.java [hunk 4]
```diff
@@ -877,7 +877,7 @@ public void createAndQueueTask(String termName, String termQName, String termGui
         String taskType = isHardDelete ? UPDATE_ENTITY_MEANINGS_ON_TERM_HARD_DELETE : UPDATE_ENTITY_MEANINGS_ON_TERM_SOFT_DELETE;
         String currentUser = RequestContext.getCurrentUser();
         Map<String, Object> taskParams = MeaningsTask.toParameters(termName, termQName, termGuid);
-        AtlasTask task = taskManagement.createTask(taskType, currentUser, taskParams);
+        AtlasTask task = taskManagement.createTask(taskType, currentUser, taskParams, termQName, termGuid);
```
Comment (mehtaanshul) [inline repository/src/main/java/org/apache/atlas/repository/store/graph/v2/AtlasEntityStoreV2.java:880]: discuss this

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v2/EntityGraphMapper.java [hunk 5]
```diff
@@ -3019,6 +3019,7 @@ public void deleteClassification(String entityGuid, String classificationName) t
                 }
 
                 if (propagateDelete) {
+                    //delete All the pending tasks of this classification vertex
```
Comment (mehtaanshul) [inline repository/src/main/java/org/apache/atlas/repository/store/graph/v2/EntityGraphMapper.java:3022]: is this a future todo item?

**Comment by nikhilbonte21** [comment]
a -> b -> c -> d -> e -> f -> g

Assume there are 2 classifications attached to vertex a, deleting edge between d-> e will create 2 tasks or only 1?

**Comment by sumandas0** [comment]
> a -> b -> c -> d -> e -> f -> g
> 
> Assume there are 2 classifications attached to vertex a, deleting edge between d-> e will create 2 REFRESH tasks or only 1?

It will create 2 task

**Comment by sumandas0** [comment]
> Dataset: a -> b -> c -> d -> e -> f -> g delete edge between d-> e first then edge between c -> d? When the Refresh task starts executing will it stop at c?

yes, because c->d link is broken, we will not able traverse beyond this.

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v2/EntityGraphMapper.java (hunk 5 above)
Comment (sumandas0) [reply repository/src/main/java/org/apache/atlas/repository/store/graph/v2/EntityGraphMapper.java]: It was just outline comment, removed it now! Ignore it.

File: repository/src/main/java/org/apache/atlas/repository/graph/GraphBackedSearchIndexer.java (hunk 1 above)
Comment (sumandas0) [reply repository/src/main/java/org/apache/atlas/repository/graph/GraphBackedSearchIndexer.java]: Yes, it should be indexed as keyword

**Comment by sumandas0** [comment]
> What will happen to the existing DELETE_ONLY_PROPAGATION tasks, will those run successfully?

I haven't modified execution part of DELETE_ONLY_PROPAGATION tasks, so all old tasks will be executed successfully now.

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v1/DeleteHandlerV1.java (hunk 3 above)
Comment (sumandas0) [reply repository/src/main/java/org/apache/atlas/repository/store/graph/v1/DeleteHandlerV1.java:1364]: Discussed it

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v1/DeleteHandlerV1.java (hunk 2 above)
Comment (sumandas0) [reply repository/src/main/java/org/apache/atlas/repository/store/graph/v1/DeleteHandlerV1.java:1322]: Removed unique and reference id , introduced classification based params

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v2/AtlasEntityStoreV2.java (hunk 4 above)
Comment (sumandas0) [reply repository/src/main/java/org/apache/atlas/repository/store/graph/v2/AtlasEntityStoreV2.java]: Removed unique and reference id , introduced classification based params

File: repository/src/main/java/org/apache/atlas/repository/graph/GraphBackedSearchIndexer.java (hunk 1 above)
Comment (sumandas0) [reply repository/src/main/java/org/apache/atlas/repository/graph/GraphBackedSearchIndexer.java]: Made those field true

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v2/EntityGraphMapper.java [hunk 6]
```diff
@@ -2865,61 +2865,54 @@ public List<String> propagateClassification(String entityGuid, String classifica
...
         }
     }
 
-    public List<String> processChunkedPropagation(List<AtlasVertex> chunkedVerticesToPropagate, AtlasVertex classificationVertex) throws AtlasBaseException{
-        try {
-            AtlasPerfMetrics.MetricRecorder metricRecorder = RequestContext.get().startMetricRecord("lockObjectsAfterTraverse");
-            List<String> impactedVerticesGuidsToLock = chunkedVerticesToPropagate.stream().map(x -> GraphHelper.getGuid(x)).collect(Collectors.toList());
-            RequestContext.get().endMetricRecord(metricRecorder);
+    public List<String> processClassificationPropagationAddition(List<AtlasVertex> verticesToPropagate, AtlasVertex classificationVertex) throws AtlasBaseException {
+        AtlasPerfMetrics.MetricRecorder chunkedPropMetricRecorder = RequestContext.get().startMetricRecord("chunkedPropagationAndNotification");
```
Comment (mehtaanshul) [inline repository/src/main/java/org/apache/atlas/repository/store/graph/v2/EntityGraphMapper.java:2877]: Change metric name to processClassificationPropagationAddition

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v2/EntityGraphMapper.java [hunk 7]
```diff
@@ -3512,24 +3527,78 @@ public void deleteClassificationOnlyPropagation(String classificationId, String
...
-    private void processClassificationDeleteOnlyPropagation(AtlasVertex currentClassificationVertex, String relationshipGuid, boolean isTermEntityEdge) throws AtlasBaseException {
-        String              classificationId                = currentClassificationVertex.getIdForDisplay();
+        AtlasVertex currentClassificationVertex = graph.getVertex(classificationId);
         String              sourceEntityId                  = getClassificationEntityGuid(currentClassificationVertex);
         AtlasVertex         sourceEntityVertex              = AtlasGraphUtilsV2.findByGuid(this.graph, sourceEntityId);
         AtlasClassification classification                  = entityRetriever.toAtlasClassification(currentClassificationVertex);
         boolean             removePropagationOnEntityDelete = getRemovePropagations(currentClassificationVertex);
 
-        if (!(isTermEntityEdge || removePropagationOnEntityDelete)) {
+        if(!removePropagationOnEntityDelete) {
```
Comment (mehtaanshul) [inline repository/src/main/java/org/apache/atlas/repository/store/graph/v2/EntityGraphMapper.java:3550]: Why no isTermEntityEdge check?

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v2/tasks/TaskUtil.java [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+
+    public TaskSearchResult findPendingTasksByClassificationId(int from, int size, String classificationId, List<String> types, List<String> excludeTypes) throws AtlasBaseException {
+        List<Map<String,Object>> mustConditions = new ArrayList<>();
+        List<Map<String,Object>> shouldConditions = new ArrayList<>();
+        List<Map<String,Object>> excludeConditions = new ArrayList<>();
+
+        if (StringUtils.isNotEmpty(classificationId))
+            mustConditions.add(getMap("term", getMap(TASK_CLASSIFICATION_ID, classificationId)));
+
+        if (CollectionUtils.isNotEmpty(types)) {
```
Comment (mehtaanshul) [inline repository/src/main/java/org/apache/atlas/repository/store/graph/v2/tasks/TaskUtil.java:36]: term query

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v2/tasks/TaskUtil.java [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+            List<Map<String, Object>> typeQueries = types.stream().map(type -> getMap("match", getMap(TASK_TYPE, type))).collect(Collectors.toList());
+            shouldConditions.addAll(typeQueries);
+        }
+
+        if(CollectionUtils.isNotEmpty(excludeConditions)) {
+            List<Map<String, Object>> excludeTypeQueries = excludeTypes.stream().map(type -> getMap("match", getMap(TASK_TYPE, type))).collect(Collectors.toList());
+            excludeConditions.addAll(excludeTypeQueries);
+        }
+
+        Map<String, Object> statusQuery = getMap("match", getMap(TASK_STATUS, TASK_STATUS_PENDING));
```
Comment (mehtaanshul) [inline repository/src/main/java/org/apache/atlas/repository/store/graph/v2/tasks/TaskUtil.java:46]: term

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v2/tasks/TaskUtil.java [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+
+        if (StringUtils.isNotEmpty(classificationId))
+            mustConditions.add(getMap("term", getMap(TASK_CLASSIFICATION_ID, classificationId)));
+
+        if (CollectionUtils.isNotEmpty(types)) {
+            List<Map<String, Object>> typeQueries = types.stream().map(type -> getMap("match", getMap(TASK_TYPE, type))).collect(Collectors.toList());
+            shouldConditions.addAll(typeQueries);
+        }
+
+        if(CollectionUtils.isNotEmpty(excludeConditions)) {
```
Comment (mehtaanshul) [inline repository/src/main/java/org/apache/atlas/repository/store/graph/v2/tasks/TaskUtil.java:41]: term

File: repository/src/main/java/org/apache/atlas/tasks/TaskManagement.java [hunk 11]
```diff
@@ -130,8 +130,8 @@ public void addFactory(TaskFactory taskFactory) {
         createTaskTypeFactoryMap(this.taskTypeFactoryMap, taskFactory);
     }
 
-    public AtlasTask createTask(String taskType, String createdBy, Map<String, Object> parameters) {
-        return this.registry.createVertex(taskType, createdBy, parameters);
+    public AtlasTask createTask(String taskType, String createdBy, Map<String, Object> parameters, String classificationId, String entityGuid) {
```
Comment (mehtaanshul) [inline repository/src/main/java/org/apache/atlas/tasks/TaskManagement.java:133]: can't we create a new method with more params and same name?

File: repository/src/main/java/org/apache/atlas/repository/graph/GraphHelper.java [hunk 12]
```diff
@@ -390,6 +390,29 @@ public static AtlasEdge getClassificationEdge(AtlasVertex entityVertex, AtlasVer
         return ret;
     }
 
+    public static boolean isClassificationAttached(AtlasVertex entityVertex, AtlasVertex classificationVertex) {
```
Comment (mehtaanshul) [inline repository/src/main/java/org/apache/atlas/repository/graph/GraphHelper.java:393]: add metric logging

Comment (mehtaanshul) [reply repository/src/main/java/org/apache/atlas/repository/graph/GraphHelper.java:393]: can we improve this logic

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v2/tasks/TaskUtil.java (hunk 8 above)
Comment (sumandas0) [reply repository/src/main/java/org/apache/atlas/repository/store/graph/v2/tasks/TaskUtil.java:36]: It is indexed as text, I've tested that match query is working perfectly here

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v2/tasks/TaskUtil.java (hunk 9 above)
Comment (sumandas0) [reply repository/src/main/java/org/apache/atlas/repository/store/graph/v2/tasks/TaskUtil.java:46]: It is indexed as text, I've tested that match query is working perfectly here

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v2/tasks/TaskUtil.java (hunk 10 above)
Comment (sumandas0) [reply repository/src/main/java/org/apache/atlas/repository/store/graph/v2/tasks/TaskUtil.java:41]: It is indexed as text, I've tested that match query is working perfectly here

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v2/EntityGraphMapper.java (hunk 6 above)
Comment (sumandas0) [reply repository/src/main/java/org/apache/atlas/repository/store/graph/v2/EntityGraphMapper.java]: Changed it

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v2/EntityGraphMapper.java (hunk 7 above)
Comment (sumandas0) [reply repository/src/main/java/org/apache/atlas/repository/store/graph/v2/EntityGraphMapper.java]: It is already checked before

File: repository/src/main/java/org/apache/atlas/tasks/TaskManagement.java (hunk 11 above)
Comment (sumandas0) [reply repository/src/main/java/org/apache/atlas/tasks/TaskManagement.java:137]: Addresses this one

**Review by mehtaanshul — APPROVED** [review]

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v2/EntityGraphMapper.java [hunk 13]
```diff
@@ -2865,61 +2865,54 @@ public List<String> propagateClassification(String entityGuid, String classifica
...
             throw new AtlasBaseException(e);
         }
     }
 
-    public List<String> processChunkedPropagation(List<AtlasVertex> chunkedVerticesToPropagate, AtlasVertex classificationVertex) throws AtlasBaseException{
-        try {
-            AtlasPerfMetrics.MetricRecorder metricRecorder = RequestContext.get().startMetricRecord("lockObjectsAfterTraverse");
-            List<String> impactedVerticesGuidsToLock = chunkedVerticesToPropagate.stream().map(x -> GraphHelper.getGuid(x)).collect(Collectors.toList());
-            RequestContext.get().endMetricRecord(metricRecorder);
+    public List<String> processClassificationPropagationAddition(List<AtlasVertex> verticesToPropagate, AtlasVertex classificationVertex) throws AtlasBaseException {
```
[the remark this answers is not in the document]
Comment (sumandas0) [reply repository/src/main/java/org/apache/atlas/repository/store/graph/v2/EntityGraphMapper.java]: Added

**Pull request merged**